### PR TITLE
treat composer service code 10 as http.status 400

### LINKED ISSUE #385: IMAGE-BUILDER-COMPOSER-10 errors should be treated as 400
The error passed from composer should be parsed, and passed along as a 400 error. As this error most likely means user input error ("Error resolving OSTree repo").

### PR DESCRIPTION
**Pull request by miyunari**
closes #385

here is my first draft. Would be nice to get already some feedback on it :)

### CODE REVIEW

File: internal/v1/server.go [hunk 1]
```diff
@@ -576,6 +576,14 @@ func (h *Handlers) ComposeImage(ctx echo.Context) error {
 			logrus.Errorf("Unable to parse composer's compose response: %v", err)
 		} else {
 			_ = httpError.SetInternal(fmt.Errorf("%s", body))
+			var serviceStat composer.ComposeStatus
```
Comment (Gundersanne) [inline internal/v1/server.go:579]: The error returned here would be of `composer.Error` type.

File: internal/v1/server_test.go [hunk 2]
```diff
@@ -677,6 +677,57 @@ func TestComposeImageErrorsWhenStatusCodeIsNotStatusCreated(t *testing.T) {
...
+		require.Equal(t, "Bearer accesstoken", r.Header.Get("Authorization"))
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		serviceStat := composer.ComposeStatus{
+			ImageStatus: composer.ImageStatus{
+				Error: &composer.ComposeStatusError{
+					Id:     10,
+					Reason: "not ok",
+				},
+			},
```
Comment (Gundersanne) [inline internal/v1/server_test.go:696]: Similar here, composer returns composer.Error, not ComposeStatus.

**Review by Gundersanne — CHANGES_REQUESTED** [review]
Sorry for the late review, the idea is correct but mostly using the wrong types here.

Composer returns an error object in case http calls fail, composer.ComposeStatus is an object that is returned when querying the status of a compose. See https://api.openshift.com/?urls.primaryName=Image%20Builder%20service.

**Comment by miyunari** [comment]
hi @Gundersanne, thanks for your feedback. Could you please have a look again? :)

File: internal/v1/server_test.go [hunk 3]
```diff
@@ -677,6 +677,55 @@ func TestComposeImageErrorsWhenStatusCodeIsNotStatusCreated(t *testing.T) {
...
+				ImageType:    ImageTypes_aws,
+				UploadRequest: UploadRequest{
+					Type: UploadTypes_aws,
+					Options: AWSUploadRequestOptions{
+						ShareWithAccounts: []string{"test-account"},
+					},
+				},
+			},
+		},
+	}
```
[the remark this answers is not in the document]
Comment (miyunari) [reply internal/v1/server_test.go]: Thanks, I added the missing parts :)

File: internal/v1/handler.go [hunk 4]
```diff
@@ -486,6 +487,20 @@ func (h *Handlers) ComposeImage(ctx echo.Context) error {
...
+
+			id, err := strconv.Atoi(serviceStat.Id)
+			if err != nil {
+				return httpError
+			}
+
+			if id == 10 {
+				httpError.Message = "Error resolving OSTree repo"
+				httpError.Code = http.StatusBadRequest
+			}
```
Comment (Gundersanne) [inline internal/v1/handler.go:503]: I don't think it's necessary to convert the id to an integer first. Just check if serviceStat.Id == "10"

**Review by Gundersanne — COMMENTED** [review]
One final comment, otherwise lgtm

File: internal/v1/handler.go (hunk 4 above)
Comment (miyunari) [reply internal/v1/handler.go]: ok, done :)

**Review by Gundersanne — DISMISSED** [review]
Thank you!

**Comment by miyunari** [comment]
Hey @Gundersanne, is there anything else I need to do? :)

**Comment by miyunari** [comment]
Just called `./tools/prepare-source.sh` again, to fix the broken test.

cc: @achilleas-k @Gundersanne

**Review by achilleas-k — APPROVED** [review]

**Comment by croissanne** [comment]
Egh, seems like a variable got marked as unrequired. Will send a separate PR.

**Comment by croissanne** [comment]
Also opened https://github.com/osbuild/image-builder/issues/467

**Review by croissanne — APPROVED** [review]

**Pull request merged**